src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java

```diff
@@ -88,7 +88,15 @@ public void activate(Map<String, Object> newConfig, Deployment deployment,
             logger.atInfo(MERGE_CONFIG_EVENT_KEY).kv(DEPLOYMENT_ID_LOG_KEY, deploymentDocument.getDeploymentId())
                     .log("All services updated");
             totallyCompleteFuture.complete(new DeploymentResult(DeploymentResult.DeploymentStatus.SUCCESSFUL, null));
-        } catch (InterruptedException | ExecutionException | ServiceUpdateException | ServiceLoadException e) {
+        } catch (InterruptedException e) {
+            // Treat interrupts distinctly: we don't want to start a rollback while the kernel is shutting down.
+            // This applies even when our failure handling policy is configured to rollback.
+            logger.atError(MERGE_CONFIG_EVENT_KEY).kv(DEPLOYMENT_ID_LOG_KEY, deploymentDocument.getDeploymentId())
+                    .setCause(e).log("Deployment interrupted: will not attempt rollback, regardless of policy");
+            totallyCompleteFuture.complete(
```

Comment:
Yes - as implemented here, a deployment status of FAILED could be sent to AWS IoT Jobs. Whereas, if we don't send any status at all (don't complete the future), then IoT Jobs should eventually set the status to TIMED_OUT on its own initiative.

For either scenario, IoT Jobs provides a [retry configuration](https://docs.aws.amazon.com/iot/latest/developerguide/jobs-configurations-details.html#job-retry-configuration).

In the deployment cancellation scenario, I don't think we can get away with not completing the future, because the DeploymentService only executes one deployment at a time. So, failing to complete the future blocks all deployments until the kernel restarts (which might not happen for a long time)

Reply:
Okay so looks like we won't land here on deployment cancellation - it won't actually cause InterruptedException only the future will be cancelled from DeploymentService/DeploymentTask. But for the shutdown case, we should ensure that we don't sometimes end up reporting or persisting status, we want it to be unchanged so can we reprocess the deployment after restart.

Reply:
Side note about Jobs timeouts/retries - timeouts are optional, and retries on timeouts/failures are not integrated in Greengrass deployments yet. And this fix needs to work for Shadow deployments too.

Reply:
Ah, you're right.  In the case of deployment cancellation, at [DefaultDeploymentTask::282](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/cbdeffa4ad846ec3aecb9fe76032232d543320b7/src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java#L282) we elect not to send an interrupt.

And in the kernel shutdown case, failing to complete the deploymentMergeFuture simply means we block at [DefaultDeploymentTask::159](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/cbdeffa4ad846ec3aecb9fe76032232d543320b7/src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java#L159) until the kernel terminates.

My only concern is, can we be 100% sure that an interrupt to DefaultActivator is always due to kernel shutdown? Because if the kernel doesn't shut down, DefaultDeploymentTask blocks indefinitely.

Reply:
`DeploymentService` waits for `DefaultDeploymentTask` here [DeploymentService::290](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/cbdeffa4ad846ec3aecb9fe76032232d543320b7/src/main/java/com/aws/greengrass/deployment/DeploymentService.java#L290). if DefaultActivator is interrupted and deployment service is still active, then result will be persisted and uploaded.

If we want to pretend that the deployment did not happen when kernel shutdown, we probably need to set DeploymentResult to null?

Also consider the following sequence of events: deployment is interrupted when part of new component versions are already started -> kernel shutdown -> deployment is failed without rollback -> kernel restart in an **offline** state, would nucleus be in a state with incomplete set of deployed components and is this desired?